src/handler.js

```diff
@@ -0,0 +1,330 @@
+    this.queue = options.queue;
+    this.scheduler = options.scheduler;
+    this.listener = options.listener;
+    this.prefix = options.prefix;
+    this.publisher = options.publisher;
+    this.validator = options.validator;
+  }
+
+  // Starts up the message handler and listens for messages
+  async start() {
+    this.listener.on('message', async (message) => {
+      await this.handleMessage(message).catch(err => {
+        console.log(`Error caught when processing message. ${err.message}. ${err.stack}`);
+      });
+    });
+    await this.listener.resume();
+  }
+
+  // Listens for Task event messages and invokes the appropriate handler
+  // for the type of message received.
+  //
+  // Only messages that contain the properly formatted routing key and contains
+  // treeherder job information in task.extra.treeherder are accepted
+  async handleMessage(message) {
+    let taskId = message.payload.status.taskId;
+    let parsedRoute = parseRouteInfo(this.prefix, taskId, message.routes);
+
+    let task = await this.queue.task(taskId);
+    validateTask(this.validator, taskId, task, TASK_TREEHERDER_SCHEMA);
+
+    switch (message.exchange) {
+      case this.queueEvents.taskPending().exchange:
+        let runId = message.payload.runId;
+        let run = message.payload.status.runs[message.payload.runId];
+        // If the task run was created for an infrastructure rerun, then resolve
+        // the previous run as retried.
+        if (taskIsRerun(runId, run)) {
+          await this.handleTaskRerun(parsedRoute, task, message.payload);
+        }
+
+        return await this.handleTaskPending(parsedRoute, task, message.payload)
+      case this.queueEvents.taskRunning().exchange:
+        return await this.handleTaskRunning(parsedRoute, task, message.payload)
+      case this.queueEvents.taskCompleted().exchange:
+        return await this.handleTaskCompleted(parsedRoute, task, message.payload)
+      case this.queueEvents.taskFailed().exchange:
+        return await this.handleTaskFailed(parsedRoute, task, message.payload)
+      case this.queueEvents.taskException().exchange:
+        return await this.handleTaskException(parsedRoute, task, message.payload)
+      default:
+        throw new Error(`Unknown exchange: ${message.exchange}`);
+    }
+
+  }
+
+  // Publishes the Treeherder job message to pulse.
+  async publishJobMessage(pushInfo, job) {
+    try {
+      await this.publisher.jobs(job, {project:pushInfo.project, destination: pushInfo.destination});
+    } catch(err) {
+      throw new Error(
+        `Could not publish job message. ${err.message}. \n` +
+        `Job: ${JSON.stringify(job, null, 4)}`);
+    }
+  }
+
+  // Builds the basic Treeherder job message that's universal for all
+  // messsage types.
+  //
+  // Specific handlers for each message type will add/remove information necessary
+  // for the type of task event..
+  buildJobMessage(pushInfo, taskId, runId, task, run) {
```

Comment:
why not take `status` and `runId`, then you don't need `run` and `taskId`..

you could even make life easier by saying that if `runId === undefined` then you are reporting the last runId, since you only have one exception to that case.